src/SystemContainer.hpp

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2024 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef GZ_SIM_SYSTEM_CONTAINER_HH_
+#define GZ_SIM_SYSTEM_CONTAINER_HH_
+
+#include <iterator>
+#include <vector>
+#include <cstddef>
+#include <functional>
+
+namespace gz
+{
+  namespace sim
+  {
+    //////////////////////////////////////////////////
+    /// This container implements a simple masked vector.
+    /// Using a masked vector for systems ensures that
+    /// when a system is deleted, its memory location stays
+    /// the same. This is important because there are a number
+    /// of places where we refer to systems by raw pointer
+    /// values.
+    ///
+    /// Operation times:
+    /// push - O(1)
+    /// removeIf - O(n)
+    /// iterate - O(n) where n is the max number of systems
+    ///  -> TODO(arjoc): It should be possible to use a cache
+    ///  to make each iterator increment have O(1) time.
+    ///  For now worst case iterator increment is O(n).
+    ///  That being said for most cases the increment
+    ///  should be O(1). Eitherways, run does not iterate
+    ///  through systems but rather ISystem* interfaces.
+    ///  The only two places we iterate over this is during
+    ///  reset and TotalByEntity in the tests.
+    template<typename T>
+    class SystemContainer
```

Comment:
Can you a put this class inside a `inline namespace GZ_SIM_VERSION_NAMESPACE`. You'll have to include `config.hh` as well.

Reply:
[d6462a2](https://github.com/gazebosim/gz-sim/pull/2232/commits/d6462a2a953f62dafc0936838276f13b6a324841)